[Unified Text Replacement] Compute the text ranges earlier in the replacement mechanism

rr-codes opened the pull request:
#### a70bd1a6f8452920e19d8312201f7f18a6d38f2a
<pre>
[Unified Text Replacement] Compute the text ranges earlier in the replacement mechanism
<a href="https://bugs.webkit.org/show_bug.cgi?id=268679">https://bugs.webkit.org/show_bug.cgi?id=268679</a>
<a href="https://rdar.apple.com/120563525">rdar://120563525</a>

Reviewed by Megan Gardner and Aditya Keerthi.

* Source/WebKit/Shared/WebUnifiedTextReplacementContextData.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::willBeginTextReplacementSession):
(WebKit::WebPageProxy::didBeginTextReplacementSession):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::willBeginTextReplacementSession):
(WebKit::WebPage::didBeginTextReplacementSession):
(WebKit::WebPage::autocorrectionContextRange):
(WebKit::WebPage::rangeSelectionContext const): Deleted.
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
(WebKit::UnifiedTextReplacementController::willBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::didBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/274115@main">https://commits.webkit.org/274115@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe15d7bf992288d9fa2021fa9df89bdf46ab99a2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37966 "Failed to checkout and rebase branch from PR 23799") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16873 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/40221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40511 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33775 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19571 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14176 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38539 "Failed to checkout and rebase branch from PR 23799") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/14219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/40221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/12414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/40221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41784 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/34445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/40221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/12949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/36417 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/14516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/40221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4922 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/WebUnifiedTextReplacementContextData.h:
```diff
@@ -25,6 +25,8 @@
 #pragma once
 
+#include <WebCore/AttributedString.h>
+#include <WebCore/CharacterRange.h>
```
Do you need these includes? I think you can get away with forward decls and that would also satiate the GTK builders

rr-codes replied:
idk why GTK is even trying to find this file, considering everywhere that uses it is behind an ifdef

aprotyas commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -41,22 +43,41 @@ UnifiedTextReplacementController::UnifiedTextReplacementController(WebPage& webP
-    auto contextRange = m_webPage->rangeSelectionContext();
+    auto* corePage = m_webPage->corePage();
+    if (!corePage)
+        return;
+
+    Ref frame = CheckedRef(corePage->focusController())->focusedOrMainFrame();
+
+    auto contextRange = m_webPage->autocorrectionContextRange();
     if (!contextRange)
         return;
 
     auto liveRange = createLiveRange(*contextRange);
 
-    ASSERT(!m_sessionRanges.contains(uuid));
+    ASSERT(!m_contextRanges.contains(uuid));
+
+    m_contextRanges.set(uuid, liveRange);
+
+    auto selectedTextRange = frame->selection().selection().firstRange();
 
-    m_sessionRanges.set(uuid, liveRange);
+    auto attributedStringFromRange = attributedString(*contextRange);
+
+    auto selectedTextCharacterRange = WebCore::characterRange(*contextRange, *selectedTextRange);
+
+    completionHandler({ WebUnifiedTextReplacementContextData { WTF::UUID { 0 }, attributedStringFromRange, selectedTextCharacterRange } });
```
Optional nit / Stylistic remark: Lots of blank lines in this method.

rr-codes replied:
I like blank lines, makes things easier to read personally

aprotyas reviewed (commented):
LGTM!

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -864,6 +863,56 @@
 
+std::optional<SimpleRange> WebPage::autocorrectionContextRange()
+{
+    // This uses the same algorithm as `WebPage::autocorrectionContext`.
```
Can any code be shared?

pxlcoder replied:
Looking at the code, it does look different enough, to the point where it can't be shared?

If that's the case, I would remove this comment and briefly describe the algorithm instead.

rr-codes committed:
[Unified Text Replacement] Compute the text ranges earlier in the replacement mechanism
https://bugs.webkit.org/show_bug.cgi?id=268679
rdar://120563525

Reviewed by Megan Gardner and Aditya Keerthi.

* Source/WebKit/Shared/WebUnifiedTextReplacementContextData.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::willBeginTextReplacementSession):
(WebKit::WebPageProxy::didBeginTextReplacementSession):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::willBeginTextReplacementSession):
(WebKit::WebPage::didBeginTextReplacementSession):
(WebKit::WebPage::autocorrectionContextRange):
(WebKit::WebPage::rangeSelectionContext const): Deleted.
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
(WebKit::UnifiedTextReplacementController::willBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::didBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/274115@main
